pkg/dataobj/explorer/service.go

```diff
@@ -0,0 +1,118 @@
+package explorer
+
+import (
+	"context"
+	"embed"
+	"encoding/json"
+	"io/fs"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/thanos-io/objstore"
+)
+
+// isDev is set via build flag -tags dev
+var isDev bool
+
+//go:embed dist
+var uiFS embed.FS
+
+type Service struct {
+	*services.BasicService
+
+	bucket   objstore.Bucket
+	logger   log.Logger
+	uiFS     fs.FS
+	devProxy *httputil.ReverseProxy
+}
+
+func New(bucket objstore.Bucket, logger log.Logger) (*Service, error) {
+	var ui fs.FS
+	var devProxy *httputil.ReverseProxy
+
+	if isDev {
+		// In dev mode, we'll proxy to the Vite dev server
+		devServerURL, err := url.Parse("http://localhost:5173")
+		if err != nil {
+			return nil, err
+		}
+		devProxy = httputil.NewSingleHostReverseProxy(devServerURL)
+		ui = os.DirFS(filepath.Join("pkg", "dataobj", "explorer", "ui", "dist")) // Fallback
```

Comment:
I think this is fine, but I wanted to let you know of the alternative approach Alloy used:

Alloy puts the proxy on the [web app itself](https://github.com/grafana/alloy/blob/45e0a1a6d2c851c9c628aa6fc0915ae7ff53a4f9/internal/web/ui/package.json#L59) instead of in the Alloy binary. So when developing the UI locally, we'd use the dev server from React and it would proxy API requests to Alloy. 

This means Alloy didn't have to know about the port for the web UI, which can help with separation of concerns a little. 

But again, this is fine :)